File: library/src/main/java/com/alamkanak/weekview/drawing/DayBackgroundDrawer.java
package com.alamkanak.weekview.drawing;

import android.graphics.Canvas;
import android.graphics.Paint;

import com.alamkanak.weekview.model.WeekViewConfig;
import com.alamkanak.weekview.utils.DateUtils;

import java.util.Calendar;

import static com.alamkanak.weekview.utils.DateUtils.isSameDay;
import static java.util.Calendar.DAY_OF_WEEK;
import static java.util.Calendar.HOUR_OF_DAY;
import static java.util.Calendar.MINUTE;
import static java.util.Calendar.SATURDAY;
import static java.util.Calendar.SUNDAY;

class DayBackgroundDrawer {

    private WeekViewConfig config;
    private WeekViewDrawingConfig drawConfig;

    DayBackgroundDrawer(WeekViewConfig config, WeekViewDrawingConfig drawConfig) {
        this.config = config;
        this.drawConfig = drawConfig;
    }

    void drawDayBackground(Calendar day, int height,
                           float startX, float startPixel, Canvas canvas) {
        Calendar today = DateUtils.today();
        boolean isToday = isSameDay(day, today);

        if (drawConfig.widthPerDay + startPixel - startX <= 0) {
            return;
        }

        float headerHeight = drawConfig.headerHeight
                + config.headerRowPadding * 2
                + drawConfig.headerMarginBottom;

        if (config.showDistinctPastFutureColor) {
            boolean isWeekend = day.get(DAY_OF_WEEK) == SATURDAY || day.get(DAY_OF_WEEK) == SUNDAY;
            boolean useWeekendColor = isWeekend && config.showDistinctWeekendColor;

            Paint pastPaint = drawConfig.getPastBackgroundPaint(useWeekendColor);
            Paint futurePaint = drawConfig.getFutureBackgroundPaint(useWeekendColor);

            float startY = headerHeight + drawConfig.timeTextHeight / 2 + drawConfig.currentOrigin.y;
            float endX = startPixel + drawConfig.widthPerDay;

            if (isToday) {
                Calendar now = Calendar.getInstance();
                float beforeNow = (now.get(HOUR_OF_DAY) + now.get(MINUTE) / 60.0f) * config.hourHeight;
                canvas.drawRect(startX, startY, endX, startY + beforeNow, pastPaint);
                canvas.drawRect(startX, startY + beforeNow, endX, height, futurePaint);
            } else if (day.before(today)) {
                canvas.drawRect(startX, startY, endX, height, pastPaint);
            } else {
                canvas.drawRect(startX, startY, endX, height, futurePaint);
            }
        } else {
            Paint todayPaint = drawConfig.getTodayBackgroundPaint(isToday);
            float top = headerHeight + drawConfig.timeTextHeight / 2;
            float right = startPixel + drawConfig.widthPerDay;
            canvas.drawRect(startX, top, right, height, todayPaint);
        }
    }

}

File: core/src/main/java/com/alamkanak/weekview/DayBackgroundDrawer.kt
package com.alamkanak.weekview

import android.graphics.Canvas
import android.graphics.Paint
import com.alamkanak.weekview.Constants.MINUTES_PER_HOUR
import java.util.Calendar
import kotlin.math.max

internal class DayBackgroundDrawer(
    private val viewState: ViewState
) : Drawer {

    override fun draw(canvas: Canvas) {
        viewState.dateRangeWithStartPixels.forEach { (date, startPixel) ->
            drawDayBackground(date, startPixel, canvas)
        }
    }

    /**
     * Draws a day's background color in the corresponding bounds.
     *
     * @param day The [Calendar] indicating the date
     * @param startPixel The x-coordinate on which to start drawing the background
     * @param canvas The [Canvas] on which to draw the background
     */
    private fun drawDayBackground(
        day: Calendar,
        startPixel: Float,
        canvas: Canvas
    ) {
        val endPixel = startPixel + viewState.widthPerDay
        val isCompletelyHiddenByTimeColumn = endPixel <= viewState.timeColumnWidth
        if (isCompletelyHiddenByTimeColumn) {
            return
        }

        val actualStartPixel = max(startPixel, viewState.timeColumnWidth)
        val height = viewState.viewHeight.toFloat()

        if (viewState.showDistinctPastFutureColor) {
            val useWeekendColor = day.isWeekend && viewState.showDistinctWeekendColor
            val pastPaint = viewState.getPastBackgroundPaint(useWeekendColor)
            val futurePaint = viewState.getFutureBackgroundPaint(useWeekendColor)

            val startY = viewState.headerHeight + viewState.currentOrigin.y
            val endX = startPixel + viewState.widthPerDay

            when {
                day.isToday -> drawPastAndFutureRect(actualStartPixel, startY, endX, pastPaint, futurePaint, height, canvas)
                day.isBeforeToday -> canvas.drawRect(actualStartPixel, startY, endX, height, pastPaint)
                else -> canvas.drawRect(actualStartPixel, startY, endX, height, futurePaint)
            }
        } else {
            val todayPaint = viewState.getDayBackgroundPaint(day.isToday)
            val right = startPixel + viewState.widthPerDay
            canvas.drawRect(actualStartPixel, viewState.headerHeight, right, height, todayPaint)
        }
    }

    private fun drawPastAndFutureRect(
        startX: Float,
        startY: Float,
        endX: Float,
        pastPaint: Paint,
        futurePaint: Paint,
        height: Float,
        canvas: Canvas
    ) {
        val now = now()
        val hour = now.hour - viewState.minHour
        val hourFraction = now.minute / MINUTES_PER_HOUR

        val beforeNow = (hour + hourFraction) * viewState.hourHeight
        canvas.drawRect(startX, startY, endX, startY + beforeNow, pastPaint)
        canvas.drawRect(startX, startY + beforeNow, endX, height, futurePaint)
    }
}
